engine/collection/ingest/engine.go

```diff
@@ -65,13 +73,44 @@ func New(
 		},
 	)
 
+	// FIFO queue for transactions
+	queue, err := fifoqueue.NewFifoQueue(
+		fifoqueue.WithCapacity(int(config.MaxMessageQueueSize)),
+		fifoqueue.WithLengthObserver(func(len int) {
+			mempoolMetrics.MempoolEntries(metrics.ResourceTransactionIngestQueue, uint(len))
+		}),
+	)
```

Comment:
general question: Any reason we seem to prefer using `fifoqueue` in the codebase for this sort of thing? Aside from maybe the fact that it allows you to register a length observer?

I think in most cases buffered channels serve this purpose just as well, if not better (allows multiple consumers to consume from it at once), and it is simpler because you don't have to manually trigger the Notifier.

Reply:
The only main benefit I see of using `fifoqueue` is that it could allow for a potentially infinitely growing queue, meaning we don't need to drop requests because of reaching capacity.

However, we eliminate this benefit if we restrict the capacity using `WithCapacity`.

Reply:
The length observer is one, and also it is more flexible in terms of using different message stores (eg. a priority queue). We initially used buffered channels when first adding a per-engine message queue, and ultimately decided against it. Another reason was that the implementation we had initially had a lot of channel management boilerplate duplicated in each engine which was difficult to reason about if you're not already quite familiar with channels. Another alternative which we talked about (maybe worth revisiting) is to have the `MessageStore` expose a channel for getting the next message to process, rather than using the notifier. Then the `MessageStore` needs to maintain a goroutine to shovel messages into that channel, and to be startable/stoppable.

Reply:
I remember a while back we discovered that the `Notifier` had a subtle bug. I don't remember the exact details, but basically if multiple messages are put into the queue concurrently, there's a race condition that could cause only a single notification to be emitted from the notifier.

So if we want to keep using this, we should either fix the Notifier or expose the channel as you suggested.

Reply:
> if multiple messages are put into the queue concurrently, there's a race condition that could cause only a single notification to be emitted from the notifier

This sounds like expected behaviour of the `Notifier` to me. If routine A and B attempt to `Notify` before routine `C` wakes to read the notification, one of the `Notify` calls from A or B will be a no-op and `C` will only observe one notification.

Reply:
This is not the problem, the problem is when there are multiple consumers, e.g `C1`, `C2`, and `C3`. They could all be idle and waiting for a notification, but with the existing `Notifier` implementation, it is possible that a producer could come and insert 20 items into the queue and call `Notify` 20 times, yet only one of the consumers would be woken up. This is undesirable, since if there are 3 workers available, we would obviously want to awaken all of them.

Reply:
Basically, the root of the issue with the existing `Notifier` implementation lies in the fact that all messages sent to a channel are delivered asynchronously, even when a consumer is already blocked on a read at the point when a send is initiated.

Therefore, in the scenario above, it's possible that when the producer sends notifications 2 to 20, notification 1 has still not been received by any receiver, even though all 3 receivers had already been blocked on a channel read prior to the first notification send.

Reply:
You're right that one `Notifier` does not work to notify multiple consumers. It is intended to send a "wake" signal to a worker when "some work" is ready. Not to send one signal per unit of work. That is why the consumer, when awoken, will process all the work available. 

I think exposing a channel as the the mechanism to read from a message queue is a better approach than what we have here.